---
title: "Analysing single-cell RNA-sequencing Pathways using Pseudo Bulk"
author: "Alexander Grentner"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analysing single-cell RNA-sequencing Pathways using Pseudo Bulk}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

## Introduction

The ReactomeGSA package is a client to the web-based Reactome Analysis
System. Essentially, it performs a gene set analysis using the latest
version of the Reactome pathway database as a backend.

This vignette shows how the ReactomeGSA package can be used to perform a
pathway analysis of cell clusters in single-cell RNA-sequencing data.

### Citation

To cite this package, use

```         
Alexander Grentner, Eliot Ragueneau, Chuqiao Gong, Adrian Prinz, Sabina Gansberger, Inigo Oyarzun, Henning Hermjakob, Johannes Griss, ReactomeGSA: new features to simplify public data reuse, Bioinformatics, Volume 40, Issue 6, June 2024, btae338, https://doi.org/10.1093/bioinformatics/btae338
```

```{r setup, include = FALSE}
    knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
    )
```

## Installation

Install the Seurat package from CRAN or github documentation from the
satijalab/Seurat is provided by
<https://satijalab.org/seurat/articles/install.html> The `ReactomeGSA`
package can be directly installed from Bioconductor:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!require(ReactomeGSA))
  BiocManager::install("ReactomeGSA")

# install the ReactomeGSA.data package for the example data
if (!require(ReactomeGSA.data))
  BiocManager::install("ReactomeGSA.data")


# install Seurat from CRAN
if(!require(Seurat))
    install.packages('Seurat')

# install to access publlic datasets√ü
if (!require(scRNAseq))
  BiocManager::install("scRNAseq")

if (!require(SingleCellExperiment))
  BiocManager::install("SingleCellExperiment")

```

### Load libraries

Load libraries from Reactome and Seurat. The Seurat package is required
to perform the subaggegration and subclustering.

```{r}
# load libraries
library(ReactomeGSA)
library(ReactomeGSA.data)
library(Seurat)
library(scRNAseq)
library(SingleCellExperiment)
```

### Prepare example dataset

```{r prepare}
# prepare example data set
SCE_OBJECT <- ZeiselBrainData()
```

Check metadata to find variable which is required by split_by variable
in the generate_pseudo_bulk_data() function

```{r seurat metadata}
metadata <- SingleCellExperiment::colData(SCE_OBJECT)
```

### Generate Pseudo bulk data

| Method        | split_by                | k parameter                          |
|------------------|----------------------|---------------------------------|
| random        | "random"                | number of random pools               |
| variable      | "variable"              | variable in metadata                 |
| subclustering | subclustering algorithm | (resolution group, comparison group) |

To set different subclustering algorithm the following are provided:

-   Louvian
-   Louvian_multilevel
-   SLM
-   Leiden

```{r bulk data}
SCE_RESULT_RANDOM <- generate_pseudo_bulk_data(SCE_OBJECT, "level1class", "random",5)  
SCE_RESULT_VARIABLE <- generate_pseudo_bulk_data(SCE_OBJECT, "level1class","variable","tissue") 
```


### Generate Metadata

The function generate_metadata() automatically generates the metadata
based on the split_by variable provided in the
generate_pseudo_bulk_data() function

```{r bulk metadata}
# generate meta data
metadata <- generate_metadata(SCE_RESULT_RANDOM)
metadata

```

```{r request analysis}
# get pathways and DEG using Reactome workflow
my_request <- ReactomeAnalysisRequest(method = "Camera")
my_request <- set_parameters(request = my_request, max_missing_values = 0.5)
my_request <- add_dataset(request = my_request, 
                          expression_values = SCE_RESULT_RANDOM, 
                          name = "Pseudo Bulk example",     # name for experiment
                          type = "rnaseq_counts",           # we expect the data to already be raw
                          comparison_factor = "Group",      # comparison for DEG;  must be string with ""
                          comparison_group_1 = "microglia",        # must be string with ""
                          comparison_group_2 = "interneurons",         # must be string with ""
                          sample_data = metadata,
                          additional_factors = NULL)

result <- perform_reactome_analysis(request = my_request)

```

### Result

Reactome provides the differentially expressed between the choosen
groups, with group 2 as reference and the regulated pathways. Both are
saved within the result object and can be accessed using the
get_result() and pathway() function.

```{r DEG result}
# get result table
result_table <- get_result(result, type="fold_changes", name = "Pseudo Bulk example")
result_table
```

```{r pathway results}
# get pathway result
result_pathways <- pathways(result)
result_pathways
```
